### PR TITLE
Allow mixing debug and release builds

### LINKED ISSUE #5013: Allow mixing release and debug builds
We've so far been strict on not allowing this, but this should be relatively easily supported. Essentially all definitions that affect our ABI like `HPX_DEBUG` should be propagated through CMake targets and the pkgconfig files. We should not "dynamically" determine whether something should be enabled or not when compiling user code. Mixing e.g. `-O0` and `-O3` flags should be ok as long as definitions like `HPX_DEBUG` are consistent across the build.

I'm not sure what how to best test that mixing builds like this works correctly though.

### PR DESCRIPTION
**Pull request by msimberg**
Fixes #5013.

Testing if this is feasible. In principle it is, but I'd appreciate feedback on if there are any pitfalls that I'm not aware of, especially on Windows. Ideally we only use `HPX_DEBUG` determine if parts of the code should be conditionally compiled/disabled/etc. However, there are a few additional macros that we use:

- `DEBUG`: I could find no reference to anything using this, does anyone know what it's doing in HPX?
- `_DEBUG`: Windows specific, added automatically in debug mode, correct? There are a couple of uses of this one. Can we replace them with `HPX_DEBUG`? (https://github.com/STEllAR-GROUP/hpx/blob/8cb1829d7ae94d1cc078a7ed1a6014c212406a09/src/runtime_distributed.cpp#L546-L551)
- `NDEBUG`: Disables `assert`. We have a few uses of `assert` which are fine, but as far as I can tell there is no reason for us to rely on this macro.
- `HPX_DISABLE_ASSERTIONS`/`BOOST_DISABLE_ASSERTIONS`: Used to disable a few assertions, but not `HPX_ASSERT`. Used together with `NDEBUG` in some cases. I'd at most leave `HPX_DISABLE_ASSERTIONS`.

Is there a use case for debug builds with assertions disabled? If yes, should it get a CMake option instead of needing to be manually added to the compile flags?

### CODE REVIEW

**Comment by hkaiser** [comment]
> * `DEBUG`: I could find no reference to anything using this, does anyone know what it's doing in HPX?

There is no reason to use this anywhere, even more as we map it onto `HPX_DEBUG` (https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25)

> * `_DEBUG`: Windows specific, added automatically in debug mode, correct? 

Yes, this auto-added in Debug builds. It is mapped onto `HPX_DEBUG`, also here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25

> There are a couple of uses of this one. Can we replace them with `HPX_DEBUG`? 

Yes.

> * `NDEBUG`: Disables `assert`. We have a few uses of `assert` which are fine, but as far as I can tell there is no reason for us to rely on this macro.

This is auto-added on windows for non-debug builds, there shouldn't be any reason for us to use it. Also, we should replace all `assert()` with `HPX_ASSERT()`. Actually, I thought we have an inspect check for this.

> * `HPX_DISABLE_ASSERTIONS`/`BOOST_DISABLE_ASSERTIONS`: Used to disable a few assertions, but not `HPX_ASSERT`. Used together with `NDEBUG` in some cases. I'd at most leave `HPX_DISABLE_ASSERTIONS`.
> 
> Is there a use case for debug builds with assertions disabled? If yes, should it get a CMake option instead of needing to be manually added to the compile flags?

I think this is a leftover from times when this macro caused `HPX_ASSERT` to fallback to `assert`. I think this can be removed altogether.

**Comment by msimberg** [comment]
> > * `DEBUG`: I could find no reference to anything using this, does anyone know what it's doing in HPX?
> 
> There is no reason to use this anywhere, even more as we map it onto `HPX_DEBUG` (https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25)
> 
> > * `_DEBUG`: Windows specific, added automatically in debug mode, correct?
> 
> Yes, this auto-added in Debug builds. It is mapped onto `HPX_DEBUG`, also here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25
> 
> > There are a couple of uses of this one. Can we replace them with `HPX_DEBUG`?
> 
> Yes.
> 
> > * `NDEBUG`: Disables `assert`. We have a few uses of `assert` which are fine, but as far as I can tell there is no reason for us to rely on this macro.
> 
> This is auto-added on windows for non-debug builds, there shouldn't be any reason for us to use it. Also, we should replace all `assert()` with `HPX_ASSERT()`. Actually, I thought we have an inspect check for this.
> 
> > * `HPX_DISABLE_ASSERTIONS`/`BOOST_DISABLE_ASSERTIONS`: Used to disable a few assertions, but not `HPX_ASSERT`. Used together with `NDEBUG` in some cases. I'd at most leave `HPX_DISABLE_ASSERTIONS`.
> > 
> > Is there a use case for debug builds with assertions disabled? If yes, should it get a CMake option instead of needing to be manually added to the compile flags?
> 
> I think this is a leftover from times when this macro caused `HPX_ASSERT` to fallback to `assert`. I think this can be removed altogether.

Perfect, thanks @hkaiser for confirming all these!

> > * `_DEBUG`: Windows specific, added automatically in debug mode, correct?
> 
> Yes, this auto-added in Debug builds. It is mapped onto `HPX_DEBUG`, also here: https://github.com/STEllAR-GROUP/hpx/blob/master/libs/core/config/include/hpx/config/debug.hpp#L19-L25

I was hoping I could remove that mapping (it's already gone on this branch), since with the mapping we can still end up with inconsistent flags (user code code compiled with `HPX_DEBUG` defined and HPX without it).

**Comment by msimberg** [comment]
> > * `NDEBUG`: Disables `assert`. We have a few uses of `assert` which are fine, but as far as I can tell there is no reason for us to rely on this macro.
> 
> This is auto-added on windows for non-debug builds, there shouldn't be any reason for us to use it. Also, we should replace all `assert()` with `HPX_ASSERT()`. Actually, I thought we have an inspect check for this.

This is not as bad as it sounds. I rechecked and we have it in the moodycamel queue header. It's already excluded from inspect. It might be worth leaving the `assert`s for easier merging of new versions, but then again it already has formatting and a few other changes to make it different so we could also change them. Don't really mind.

The other place where we use `assert` is for assertions in device code, which I also think is fine.

File: cmake/HPX_CompilerFlagsTargets.cmake [hunk 1]
```diff
@@ -21,6 +21,13 @@ target_compile_definitions(
   hpx_public_flags INTERFACE $<$<CONFIG:Debug>:HPX_DEBUG>
 )
 
+target_compile_definitions(
+  hpx_private_flags
+  INTERFACE $<$<CONFIG:MinSizeRel>:NDEBUG> hpx_private_flags
+  INTERFACE $<$<CONFIG:Release>:NDEBUG> hpx_private_flags
```
Comment (hkaiser) [inline cmake/HPX_CompilerFlagsTargets.cmake:27]: Do we need the private flags twice?

File: cmake/templates/HPXMacros.cmake.in [hunk 2]
```diff
@@ -95,25 +95,3 @@ function(hpx_check_allocator_compatibility)
...
-    else()
-      hpx_error("CMAKE_BUILD_TYPE does not match ${HPX_BUILD_TYPE}: "
-        "this project uses '${CMAKE_BUILD_TYPE}', HPX uses '${HPX_BUILD_TYPE}'. "
-        "Please add -DCMAKE_BUILD_TYPE=${HPX_BUILD_TYPE} to your cmake command "
-        "or add -DHPX_IGNORE_CMAKE_BUILD_TYPE_COMPATIBILITY=ON to ignore this "
-        "check (not recommended; HPX is not ABI compatible between release and "
-        "debug modes).")
-    endif()
-  endif()
-endfunction()
```
Comment (hkaiser) [inline cmake/templates/HPXMacros.cmake.in:119]: Should we at least leave a warning?

**Review by hkaiser — COMMENTED** [review]
Should we add CI configurations making sure using a Debug version from a release build and v.v. is actually functional? I'd like to see that for all major platforms (at least Linux and Windows).

File: cmake/HPX_CompilerFlagsTargets.cmake (hunk 1 above)
Comment (msimberg) [reply cmake/HPX_CompilerFlagsTargets.cmake:27]: That's a bug, it shouldn't be there at all. Good catch!

File: cmake/templates/HPXMacros.cmake.in (hunk 2 above)
Comment (msimberg) [reply cmake/templates/HPXMacros.cmake.in:119]: I'm not sure. My thinking was that this is now fully the users responsibility and mixing build types is a normal thing to do, but I can add back a warning if you feel it's appropriate.

**Comment by msimberg** [comment]
> Should we add CI configurations making sure using a Debug version from a release build and v.v. is actually functional? I'd like to see that for all major platforms (at least Linux and Windows).

That's my last concern. The build unit tests now build the hello world component example in four different build types. It's not exhaustive but it's a start. A more comprehensive test would be to make use of the feature to build tests against an installed HPX. Let's at the very least wait until we can get the windows builders running again.

File: cmake/templates/HPXMacros.cmake.in (hunk 2 above)
Comment (hkaiser) [reply cmake/templates/HPXMacros.cmake.in:119]: We don't need a warning if we can ensure that everything works without issues.

**Comment by msimberg** [comment]
retest lsu

**Comment by hkaiser** [comment]
This issue is not directly connected, but it's related: https://github.com/STEllAR-GROUP/phylanx/issues/1311

File: cmake/templates/HPXMacros.cmake.in [hunk 3]
```diff
@@ -95,25 +95,3 @@ function(hpx_check_allocator_compatibility)
       "informational purposes to dependent projects and should not be changed.")
   endif()
 endfunction()
-
-# Check we are using the same build type as HPX
-function(hpx_check_cmake_build_type)
-  get_property(_GENERATOR_IS_MULTI_CONFIG GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
-  if (NOT _GENERATOR_IS_MULTI_CONFIG AND
-      NOT "${HPX_BUILD_TYPE}" STREQUAL "${CMAKE_BUILD_TYPE}" AND
```
Comment (msimberg) [inline cmake/templates/HPXMacros.cmake.in:103]: @hkaiser, @K-ballo, we used to skip checking that build types match for Visual Studio projects anyway. Is there anything we need to or can do to check that, or is that handled somehow automatically by Visual Studio?

Comment (hkaiser) [reply cmake/templates/HPXMacros.cmake.in:103]: @msimberg I'm not sure we can do anything about it at configuration time as the generated VS projects are multi-target. The mismatch should be caught by the versioning symbols we generate, though (when building applications).

**Comment by msimberg** [comment]
@hkaiser any idea what I might've missed here: https://github.com/STEllAR-GROUP/hpx/pull/5099/checks?check_run_id=1870902759#step:5:1476?

**Comment by hkaiser** [comment]
> @hkaiser any idea what I might've missed here: https://github.com/STEllAR-GROUP/hpx/pull/5099/checks?check_run_id=1870902759#step:5:1476?

Let me try to reproduce this locally.

**Comment by hkaiser** [comment]
> > @hkaiser any idea what I might've missed here: https://github.com/STEllAR-GROUP/hpx/pull/5099/checks?check_run_id=1870902759#step:5:1476?
> 
> Let me try to reproduce this locally.

Looks like that the same commands just work fine for me locally. So it has to be some resource restriction imposed by github. We could try to use `-verbosity:detailed` (instead of the `-verbosity:minimal`) and hope to see more information.

**Comment by msimberg** [comment]
> > > @hkaiser any idea what I might've missed here: https://github.com/STEllAR-GROUP/hpx/pull/5099/checks?check_run_id=1870902759#step:5:1476?
> > 
> > 
> > Let me try to reproduce this locally.
> 
> Looks like that the same commands just work fine for me locally. So it has to be some resource restriction imposed by github. We could try to use `-verbosity:detailed` (instead of the `-verbosity:minimal`) and hope to see more information.

Thanks for checking. I'll try that out.

**Comment by hkaiser** [comment]
@msimberg I pushed some changes to this branch that work around an MSVC issue that suddenly manifested itself on this branch. Let's see if that fixes the problems for the Windows builders,

**Comment by hkaiser** [comment]
FWIW, I think the remaining issue is caused by the additional targets you added (`cmake_debug_build_dir_targets_test` and friends).

File: .github/workflows/windows_release.yml [hunk 4]
```diff
@@ -38,19 +38,28 @@ jobs:
...
           cmake --build build --config Release \
-            --target cmake_install_dir_targets_test.make_compile \
-              cmake_install_dir_macros_test.make_compile \
+            --target \
+              cmake_release_install_dir_targets_test.make_build_dir \
+              cmake_release_install_dir_targets_test.make_configure \
+              cmake_release_install_dir_targets_test.make_compile   \
+              cmake_release_install_dir_macros_test.make_build_dir \
+              cmake_release_install_dir_macros_test.make_configure \
+              cmake_release_install_dir_macros_test.make_compile   \
```
Comment (msimberg) [inline .github/workflows/windows_release.yml:62]: Grr, sorry @hkaiser for wasting your time with this. Thanks for catching this!

**Comment by hkaiser** [comment]
@msimberg this looks like it's good to go.

**Comment by msimberg** [comment]
retest lsu

**Comment by msimberg** [comment]
> @msimberg this looks like it's good to go.

Thanks you @hkaiser for helping with this!

**Comment by rbyshko** [comment]
I have installed hpx 1.6.0 using vcpkg and tried to build a "Hello world" program as in Quick Start from the documentation. Unfortunately, I still get the error this issue claims to have fixed.

With vcpkg I have also installed the current master (4d748091cc937b9eb61b7b42562191e1e2b638e3), but the error is still there.

```
terminate called after throwing an instance of 'hpx::detail::exception_with_info<hpx::exception>'
  what():  failed to insert base_set_event_action into typename to id registry.: HPX(invalid_status)

```
Let me know if I can provide more information to get this fixed.

**Comment by msimberg** [comment]
Since you mention vcpkg I'm guessing you're on Windows, is that correct? If that's the case this PR does not allow mixing build types on Windows. I don't know if we can do more to detect these situations, any help with that would definitely be appreciated.

**Comment by rbyshko** [comment]
No, I am using vcpkg on Linux, Ubuntu 20.04 to be exact. vcpkg installs release and debug versions of the hpx. However they go into separate directories. CMake seems to handle it correctly:

image

**Comment by msimberg** [comment]
> No, I am using vcpkg on Linux, Ubuntu 20.04 to be exact. vcpkg installs release and debug versions of the hpx. However they go into separate directories. CMake seems to handle it correctly:
> 
> image

Ok, thanks for the details! That does sound like something I've missed in that case, I'll try to have a look at this soon. As you can see this is rather new, and our testing (clearly) isn't thorough enough for it.

**Comment by hkaiser** [comment]
> I have installed hpx 1.6.0 using vcpkg and tried to build a "Hello world" program as in Quick Start from the documentation. Unfortunately, I still get the error this issue claims to have fixed.
> 
> With vcpkg I have also installed the current master ([4d74809](https://github.com/STEllAR-GROUP/hpx/commit/4d748091cc937b9eb61b7b42562191e1e2b638e3)), but the error is still there.
> 
> ```
> terminate called after throwing an instance of 'hpx::detail::exception_with_info<hpx::exception>'
>   what():  failed to insert base_set_event_action into typename to id registry.: HPX(invalid_status)
> ```
> 
> Let me know if I can provide more information to get this fixed.

This error message indicates that both versions, release and debug, have been loaded. Apparently we still miss something as it shouldn't actually happen. The plugin API we use relies on different function names for release and debug, so there is at least a potential way of preventing to load a mismatched version.

**Comment by msimberg** [comment]
@rbyshko just for completeness, could you provide the code and CMakeLists.txt that you used? And are you seeing this problem with user code in release mode and HPX in debug mode, vice versa, or both?

**Comment by rbyshko** [comment]
I am experiencing this error in both release and debug builds.

CMakeLists.txt
```
cmake_minimum_required(VERSION 3.16)

set(CMAKE_TOOLCHAIN_FILE ${CMAKE_CURRENT_LIST_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake)

project(HPX)

find_package(HPX CONFIG REQUIRED)

add_executable(hpx_hello hpx_hello.cpp)
target_link_libraries(hpx_hello PRIVATE HPX::hpx HPX::wrap_main HPX::iostreams_component)
```

hpx_hello.cpp
```
#include <hpx/hpx_main.hpp>
#include <hpx/iostream.hpp>

int main()
{
    // Say hello to the world!
    hpx::cout << "Hello World!\n" << hpx::flush;
    return 0;
}

```

**Pull request merged**